docs/compactor.md

```diff
@@ -0,0 +1,93 @@
+
+**Proposed Directory Structure**
+```yaml
+Backup :
+    - Full-Snapshot-revision-1
+    - Incremental-Snapshot-revision-6-1 (Format- Incremental-Snapshot-revision-<Revision Number After Last Operation>-<Revision Number Before First Operation>)
+    - Incremental-Snapshot-revision-12-6
+    - Incremental-Snapshot-revision-14-12
+    - Incremental-Snapshot-revision-18-14
+    - Full-Snapshot-revision-18 (Compacted)
+    - Incremental-Snapshot-revision-22-18
+    - Incremental-Snapshot-revision-27-22
+    - Incremental-Snapshot-revision-32-27
+    - Incremental-Snapshot-revision-38-32
+    - Full-Snapshot-revision-38 (Compacted)
+    - Incremental-Snapshot-revision-41-38
+    - Incremental-Snapshot-revision-43-41
+    - Incremental-Snapshot-revision-48-43
+    - Incremental-Snapshot-revision-52-48
+```
+
+3. A meta-data file might need to be added in the Backup directory to mark which compacted (or full) snapshot is ready to be used as the base for restoration. Restorer would study the meta-data file and pickup the ready full snapshot from there and apply all the delta snapshots having revision number greater than that. So, with the backup directory as above, if the content of metadata file is like following:
+
+* Full-Snapshot-revision-1
+* Full-Snapshot-revision-18
+
+Then, during restoration Full-Snapshot-revision-18, and all the delta snaphots from Incremental-Snapshot-revision-22-18 to Incremental-Snapshot-revision-52-48 will be applied. **Restorer would restore from latest ready full snapshot and all subsequent delta snapshots.** As Full-Snapshot-revision-38 is not present in the metadata file, it would not be considered as ready yet.
+
+Another approach to make sure that restorer only picks up snapshots which are ready, full snapshots can be named with extension `.part` until the snapshot is ready. When the snapshot is ready, `.part` extension can be removed from it's name. Restorer would avoid any file having extension as `.part`.  
+
+So, if Full-Snapshot-revision-38 is not ready in the above example, it can be named as `Full-Snapshot-revision-38.part`. Restorer will apply Full-Snapshot-revision-18 and all the delta snaphots from Incremental-Snapshot-revision-22-18 to Incremental-Snapshot-revision-52-48. **Restorer would restore from latest ready full snapshot and all subsequent delta snapshots.**
+
+This option is preferred as it avoids introducing a metadata file. But it would work only if renaming to remove the `.part` extension (upon completion of upload) is an atomic operation in all the supported snapstores. This needs to be evaluated.
+
+#### Backward Compatibility
```

Comment:
I am wary of all the effort with backward compatibility. Isn't there a simple and robust way around that by detecting and rewriting the backup structure?

I assume, the main problem are the revisions we are now encoding into the names? How about "faking it"?
- We leave the old structure as is until a future release for human operators / emergency measures
- We take the latest full snapshot and then all its delta snapshots and copy them over (with the marker file approach from above, just to be on the safe side)
- As revisions, we fake it by starting at 0 and incrementing by 1, i.e. the full snapshot is 0, the first delta is 1, the next 2, etc.

The resulting revisions in the filenames will be far too small, but who cares? The algorithm picks the highest full snapshot and then all consecutive deltas and they would all be in order. The next full/incremental snapshots would carry the proper revisions.

Note: We discussed the need for revisions only in the context of clustered ETCD. The migration of the old backup structure will be completed long before we can ship clustered ETCD. Only then we need true revisions, before that only ordering.

WDYT? This way, even if ETCD dies on us, migration can be executed before restore and restore can work only from new backup structure.

Reply:
The main problem is not only the naming parts. Actually that is not much of problem. We are currently saving full snapshots and subsequent delta snapshots till the next full snapshots together into separate directory. The main problem is that we are planning to scrap this directory structure and bring all the full snapshots and delta snapshots under one directory

Reply:
@abdasgupta Can you please respond also to the other comments and also expand your explanation above?